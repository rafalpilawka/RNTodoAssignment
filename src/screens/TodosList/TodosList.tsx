import React, {FC, ReactElement, useEffect} from 'react';
import {ActivityIndicator, View} from 'react-native';
import {EmptyListComponent} from 'src/components/EmptyListComponent/EmptyListComponent';
import {ScrollWrapper} from 'src/components/ScrollWrapper/ScrollWrapper';
import {colors} from 'src/constants/platformSpecificColors';
import {useSelector, useDispatch} from 'react-redux';
import {
  getTodoListAction,
  removeTodoAction,
  toggleTodoAction,
} from 'src/store/todos/todos.actions';
import {
  selectList,
  selectLoading,
} from 'src/store/todos/todos.selector';
import {Header} from 'src/components/Header/Header';
import {Todo} from 'src/components/Todo/Todo';
import {ListType, TodoType} from 'src/store/todos/todos.types';

export type TodoListPropsT = {
  listPointer: ListType;
  back(): void;
};

export const TodosList: FC<TodoListPropsT> = ({
  listPointer,
  back,
}): ReactElement => {
  const mainLoader = useSelector(selectLoading);
  const list = useSelector(selectList);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getTodoListAction(listPointer.id));
  }, [dispatch]);

  const _removeTodo = (id: string) => {
    dispatch(removeTodoAction({id, listId: listPointer.id}));
  };

  const _toggleTodo = (id: string) => {
    dispatch(toggleTodoAction({id, listId: listPointer.id}));
  };

  const _renderList = () =>
    list.length ? (
      (list as Array<TodoType>).map((el) => {
        return (
          <Todo
            item={el}
            key={el.id}
            removeHandler={_removeTodo}
            toggleHandler={_toggleTodo}
          />
        );
      })
    ) : (
      <EmptyListComponent variant={'TODOS'} />
    );

  return (
    <>
      <Header back={back} listId={listPointer.name} />
      {mainLoader ? (
        <View
          style={{
            flexGrow: 1,
            width: '100%',
            height: '100%',
            position: 'absolute',
            justifyContent: 'center',
            alignItems: 'center',
          }}>
          <ActivityIndicator animating size={'large'} color={colors!.color} />
        </View>
      ) : (
        <ScrollWrapper>{_renderList()}</ScrollWrapper>
      )}
    </>
  );
};
