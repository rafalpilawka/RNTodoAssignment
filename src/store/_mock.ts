export const mockTodos = [
  {id: 'aaa', title: 'Add those todos', completed: false},
  {id: 'aaa2233', title: 'Delete second', completed: false},
  {id: 'bbnjhjhjhjh', title: 'Add third one', completed: true},
];

export const mockLists = [
  {name: 'List 1', id: '11242342144'},
  {name: 'list 2', id: '11242342122'},
  {name: 'list 3', id: '1124234211'},
];

export const mockStore = {
  index: mockLists,
  todos: mockTodos,
  loading: false,
  errors: undefined,
  idLoader: undefined,
};

export const mockInitialStore = {
  index: [],
  todos: [],
  loading: false,
  errors: undefined,
  idLoader: undefined,
};
