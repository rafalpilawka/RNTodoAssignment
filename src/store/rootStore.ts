import {all, call} from 'redux-saga/effects';
import {applyMiddleware, combineReducers, createStore, Reducer} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import todosReducer from 'src/store/todos/todos.reducer';
import {watchTodosSaga} from 'src/store/todos/todos.saga';

const sagaMiddleware = createSagaMiddleware();

function* rootSaga() {
  yield all([
    call(watchTodosSaga),
    //OTHER WATCHERS GOES HERE
  ]);
}

// const store: Reducer<RootState> = combineReducers<RootState>({
//   todos: todosReducer,
// });

const rootStore = (state: any, action: any) => {
  return todosReducer(state, action);
};

// export interface RootState {
//   todos: ReduxState;
// }

export type RootState = ReturnType<typeof todosReducer>;

export default createStore(
  rootStore,
  composeWithDevTools(applyMiddleware(sagaMiddleware)),
);
sagaMiddleware.run(rootSaga);
