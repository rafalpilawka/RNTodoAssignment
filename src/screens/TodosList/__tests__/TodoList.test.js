import React from 'react';
import 'react-native';
import {TodosList} from 'src/screens/TodosList/TodosList';
import renderer from 'react-test-renderer';
import {mockStore} from 'src/store/_mock';

jest.mock('react-redux', () => ({
  useDispatch: () => () => jest.fn(),
  useSelector: (selector) => selector(mockStore),
}));
const back = jest.fn();
const pointer ={name: 'List 1', id: '11242342144'}

describe('ListsIndex', () => {
  describe('Render snapshot', () => {
    it('should match to snapshot', () => {
      const component = renderer
        .create(<TodosList back={back} listPointer={pointer}/>)
        .toJSON();
      expect(component).toMatchSnapshot();
    });
  });
});
