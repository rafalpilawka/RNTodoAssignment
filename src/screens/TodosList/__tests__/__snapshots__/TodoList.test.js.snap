// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsIndex Render snapshot should match to snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "alignItems": "center",
          "backgroundColor": Object {
            "semantic": Array [
              "systemBlue",
            ],
          },
          "borderBottomWidth": 1,
          "height": 100,
          "justifyContent": "space-between",
          "marginBottom": 30,
          "paddingHorizontal": 40,
          "width": "100%",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontFamily": "Damascus",
              "fontSize": 40,
              "fontWeight": "bold",
            },
            Object {
              "paddingBottom": 12,
            },
          ]
        }
      >
        &lt;
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Damascus",
          "fontSize": 40,
          "fontWeight": "bold",
        }
      }
    >
      LIST 1
    </Text>
    <View />
  </View>,
  <RCTScrollView
    bounces={true}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "marginBottom": 20,
        "marginHorizontal": 40,
        "paddingBottom": 40,
      }
    }
    keyboardShouldPersistTaps="always"
    showsVerticalScrollIndicator={false}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "black",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 160,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          pointerEvents="auto"
        >
          <RNCCheckbox
            forwardedRef={null}
            onValueChange={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "height": 32,
                  "width": 32,
                },
                undefined,
              ]
            }
            value={false}
          />
        </View>
        <Text
          style={
            Object {
              "flex": 1,
              "fontSize": 32,
              "paddingHorizontal": 40,
            }
          }
        >
          Add those todos
        </Text>
        <View
          accessible={true}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "red",
              "borderRadius": 100,
              "height": 60,
              "justifyContent": "center",
              "width": 60,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            x
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "black",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 160,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          pointerEvents="auto"
        >
          <RNCCheckbox
            forwardedRef={null}
            onValueChange={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "height": 32,
                  "width": 32,
                },
                undefined,
              ]
            }
            value={false}
          />
        </View>
        <Text
          style={
            Object {
              "flex": 1,
              "fontSize": 32,
              "paddingHorizontal": 40,
            }
          }
        >
          Delete second
        </Text>
        <View
          accessible={true}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "red",
              "borderRadius": 100,
              "height": 60,
              "justifyContent": "center",
              "width": 60,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            x
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "black",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 160,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          pointerEvents="auto"
        >
          <RNCCheckbox
            forwardedRef={null}
            onValueChange={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "height": 32,
                  "width": 32,
                },
                undefined,
              ]
            }
            value={true}
          />
        </View>
        <Text
          style={
            Object {
              "flex": 1,
              "fontSize": 32,
              "paddingHorizontal": 40,
            }
          }
        >
          Add third one
        </Text>
        <View
          accessible={true}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "red",
              "borderRadius": 100,
              "height": 60,
              "justifyContent": "center",
              "width": 60,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            x
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
